Modules/blake2module.c

```diff
@@ -0,0 +1,928 @@
+ * 64 bits so we loop in <4gig chunks when needed. */
+
+#if PY_SSIZE_T_MAX > UINT32_MAX
+#define HACL_UPDATE_LOOP(update,state,buf,len) \
+  while (len > UINT32_MAX) { \
+    update(state, buf, UINT32_MAX); \
+    len -= UINT32_MAX; \
+    buf += UINT32_MAX; \
+  }
+#else
+#define HACL_UPDATE_LOOP(update,state,buf,len)
+#endif
+
+#define HACL_UPDATE(update,state,buf,len) do { \
+  /* Note: we explicitly ignore the error code on the basis that it would take >
+   * 1 billion years to overflow the maximum admissible length for SHA2-256
+   * (namely, 2^61-1 bytes). */ \
+  HACL_UPDATE_LOOP(update,state,buf,len) \
+  /* Cast to uint32_t is safe: len <= UINT32_MAX at this point. */ \
+  update(state, buf, (uint32_t) len); \
+} while (0)
+
+static void update(Blake2Object *self, uint8_t *buf, Py_ssize_t len) {
+    switch (self->impl) {
+      // These need to be ifdef'd out otherwise it's an unresolved symbol at
+      // link-time.
+#ifdef HACL_CAN_COMPILE_SIMD256
+        case Blake2b_256:
+            HACL_UPDATE(Hacl_Hash_Blake2b_Simd256_update,self->blake2b_256_state, buf, len);
+            return;
+#endif
+#ifdef HACL_CAN_COMPILE_SIMD128
+        case Blake2s_128:
+            HACL_UPDATE(Hacl_Hash_Blake2s_Simd128_update,self->blake2s_128_state, buf, len);
+            return;
+#endif
+        case Blake2b:
+            HACL_UPDATE(Hacl_Hash_Blake2b_update,self->blake2b_state, buf, len);
+            return;
+        case Blake2s:
+            HACL_UPDATE(Hacl_Hash_Blake2s_update,self->blake2s_state, buf, len);
+            return;
+        default:
+            Py_UNREACHABLE();
+    }
+}
+
+static PyObject *
+py_blake2b_or_s_new(PyTypeObject *type, PyObject *data, int digest_size,
+                    Py_buffer *key, Py_buffer *salt, Py_buffer *person,
+                    int fanout, int depth, unsigned long leaf_size,
+                    unsigned long long node_offset, int node_depth,
+                    int inner_size, int last_node, int usedforsecurity)
+
+{
+    Blake2Object *self = NULL;
+    Py_buffer buf;
+
+    self = new_Blake2Object(type);
+    if (self == NULL) {
+        goto error;
+    }
+
+    self->impl = type_to_impl(type);
+
+    // Using Blake2b because we statically know that these are greater than the
+    // Blake2s sizes -- this avoids a VLA.
+    uint8_t salt_[HACL_HASH_BLAKE2B_SALT_BYTES] = { 0 };
+    uint8_t personal_[HACL_HASH_BLAKE2B_PERSONAL_BYTES] = { 0 };
+
+    /* Validate digest size. */
+    if (digest_size <= 0 ||
+        (unsigned) digest_size > (is_blake2b(self->impl) ? HACL_HASH_BLAKE2B_OUT_BYTES : HACL_HASH_BLAKE2S_OUT_BYTES))
+    {
+        PyErr_Format(PyExc_ValueError,
+                "digest_size for %s must be between 1 and %d bytes, here it is %d",
+                is_blake2b(self->impl) ? "Blake2b" : "Blake2s",
+                is_blake2b(self->impl) ? HACL_HASH_BLAKE2B_OUT_BYTES : HACL_HASH_BLAKE2S_OUT_BYTES,
+                digest_size);
+        goto error;
+    }
```

Comment:
You _could_ implement all the validators in this function as custom Argument Clinic converters. It would render this "new" helper a lot more readable. This could also be added as an improvement in a follow-up PR.

Reply:
I am all for this -- I just didn't know there was such as thing. Is there documentation somewhere? Happy to do this in a separate PR if that's ok.

Reply:
Check [the Argument Clinic section in the devguide](https://devguide.python.org/development-tools/clinic/). Also, there are examples of such converters in the stdlib (`git grep 'python input' Modules/` should get you started).